code/WorkInProgress/AzrunStuff.dm

```diff
@@ -58,4 +58,352 @@
+	var/static/list/possible_transformations
+
+	New(teg)
+		. = ..()
+		generator = teg
+		if(!possible_transformations)
+			possible_transformations = list()
+			for(var/T in childrentypesof(/datum/teg_transformation))
+				var/datum/teg_transformation/TT = new T
+				possible_transformations += TT
+
+	disposing()
+		. = ..()
+		generator = null
+
+	proc/check_reagent_transformation()
+		for(var/datum/teg_transformation/T as() in possible_transformations)
+			if(generator.active_form?.type == T.type) continue // Skip current form
+
+			var/reagents_present = length(T.required_reagents)
+			for(var/R as() in T.required_reagents)
+				if(generator.circ1.reagents.get_reagent_amount(R) + generator.circ2.reagents.get_reagent_amount(R) >= T.required_reagents[R])
+				else
+					reagents_present = FALSE
+					break
+
+			if(reagents_present)
+				SPAWN_DBG(0)
+					if(generator.active_form)
+						generator.active_form.on_revert()
+					generator.active_form = new T.type
+					generator.active_form.on_transform(generator)
+				return
+
+	proc/check_material_transformation()
+		// Transform when a matsci semiconductor is inserted and the material differs the material
+		// from the TEG.  Transformation requires the semiconductor fully back in place and energy
+		// is present to activate NANITES!
+		if(!generator.active_form || generator.active_form.type == /datum/teg_transformation/matsci)
+			if(generator.semiconductor?.material && ((src.generator.semiconductor.material.mat_id != src.generator.material?.mat_id) || !src.generator.material))
+				if(src.generator.semiconductor_state == 0 && src.generator.powered())
+					SPAWN_DBG(15.0)
+						src.generator.use_power(500)
+						elecflash(src.generator)
+						src.generator.visible_message("<span class='alert'>[src.generator] is suddenly engulfed in a swarm of nanites!</span>")
+						var/nanite_overlay = image('icons/misc/critter.dmi', "nanites")
+						generator.UpdateOverlays(nanite_overlay,"transform")
+						generator.circ1.UpdateOverlays(nanite_overlay,"transform")
+						generator.circ2.UpdateOverlays(nanite_overlay,"transform")
+						sleep(rand(15,25))
+						if(generator.active_form)
+							generator.active_form.on_revert()
+						generator.active_form = new /datum/teg_transformation/matsci
+						generator.active_form.material = generator.semiconductor.material.mat_id
+						generator.active_form.on_transform(generator)
+						sleep(rand(15,25))
+						src.generator.visible_message("<span class='alert'>The swarm of nanites disappears back into \the [src.generator].</span>")
+						generator.UpdateOverlays(null,"transform")
+						generator.circ1.UpdateOverlays(null,"transform")
+						generator.circ2.UpdateOverlays(null,"transform")
+
+
+ABSTRACT_TYPE(/datum/teg_transformation)
+datum
+	teg_transformation
+		var/name = null
+		var/id = null
+		var/audio_clip
+		var/visible_msg
+		var/audible_msg
+		var/teg_overlay
+		var/circulator_overlay
+		var/material
+		var/list/required_reagents
```

Comment:
what can be in this list

Reply:
Added dmdoc friendly comment commenting what it is because it is a list of strings!